File: clojure/binary-search-tree/src/binary_search_tree.clj
(ns binary-search-tree)

(def value :value)
(def left :left)
(def right :right)

(defn singleton [val] {:value val})

(defn insert [val bst]
  (cond (nil? bst) (singleton val)
        (<= val (value bst)) (assoc bst :left (insert val (left bst)))
        :else (assoc bst :right (insert val (right bst)))))

(defn from-list [coll]
  (reduce (fn [bst val] (insert val bst))
          (singleton (first coll))
          (rest coll)))

(defn to-list [bst]
  (if (nil? bst) []
      (concat (to-list (left bst))
              (list (value bst))
              (to-list (right bst)))))

File: go/binary-search-tree/binary_search_tree.go
package binarysearchtree

type SearchTreeData struct {
	data        int
	left, right *SearchTreeData
}

func Bst(n int) *SearchTreeData {
	return &SearchTreeData{data: n}
}

func (b *SearchTreeData) Insert(n int) {
	if n > b.data {
		if b.right != nil {
			b.right.Insert(n)
		} else {
			b.right = Bst(n)
		}
	} else {
		if b.left != nil {
			b.left.Insert(n)
		} else {
			b.left = Bst(n)
		}
	}
}

func (b *SearchTreeData) MapInt(f func(int) int) []int {
	var ints = []int{}
	if b.left != nil {
		ints = append(ints, b.left.MapInt(f)...)
	}
	ints = append(ints, f(b.data))
	if b.right != nil {
		ints = append(ints, b.right.MapInt(f)...)
	}
	return ints
}

func (b *SearchTreeData) MapString(f func(int) string) []string {
	var strings = []string{}
	if b.left != nil {
		strings = append(strings, b.left.MapString(f)...)
	}
	strings = append(strings, f(b.data))
	if b.right != nil {
		strings = append(strings, b.right.MapString(f)...)
	}
	return strings
}
